### ISSUE TITLE
Provide a distinction between URI, URL and URN

### ISSUE THREAD

**Issue by nikita-volkov** [Soostone/uri-bytestring#18]
Please see
* https://www.reddit.com/r/haskell/comments/3fzrpm/looking_for_a_proper_url_library/ctttrud
* http://stackoverflow.com/questions/176264/what-is-the-difference-between-a-uri-a-url-and-a-urn/1984274#1984274

**Comment by MichaelXavier** [comment]
Makes sense. Unfortunately this will require a breaking release (which we're willing to do at this point) because I regrettably chose "URI" instead of "URL". So the steps to get this done are:

1. Implement a URN parser. Even though the RFC seems kind of vague on it, I've come across a BNF grammar for it so it should be pretty easy.
2. Rename what's currently URI to URL
3. Add a sum type URI that spans URN and URL.
3. Expose toplevel parsers for URI in case you want to actually parse something that could be either.
